Add hydra

vict0rsch opened the pull request:
Adding a cool feature: Hydra to map command-line args to configuration:

now `python train.py data.loaders.num_workers=10 gen.encoder.n_res=128` will do exactly what you'd expect it to do

see [https://hydra.cc](https://hydra.cc) (v0.11)

I'll explain how it works when I'm sure it's all good

vict0rsch commented on train.py:
```diff
@@ -63,6 +63,7 @@ def main(opts):
             Path(opts.output_path).mkdir()
 
         # Save config file
+        # TODO what if resuming? re-dump?
```
FYI

tianyu-z commented on train.py:
```diff
@@ -1,53 +1,16 @@
+import hydra
+import yaml
 from addict import Dict
+from comet_ml import Experiment
```
I remembered that the `from comet_ml import Experiment` needs to be the first line. Correct me if I am wrong.

vict0rsch replied:
It does not need to be first but it does need to be before `torch` and other tensor stuff that's true

alexrey88 commented on omnigan/utils.py:
```diff
@@ -29,32 +49,38 @@ def load_opts(path=None, default=None):
+            overriding_opts = yaml.safe_load(f)
 
-    default_opts.update(overriding_opts)
+    opts = Dict(merge(overriding_opts, default_opts))
```
Just wondering why you created a merge() function instead of using dict.update()? :)

vict0rsch replied:
because for some reason it wasn't recursive so:

```
{a: {b:c, d:e}}.update({a: {b:c, f:g}})
```

would give `{a: {b:c, f:g}` instead of `{a: {b:c, d:e, f:g}}`

vict0rsch replied:
or something like that :p

alexrey88 replied:
Oh I see, thanks!

sashavor commented:
Looks like there is a conflict that should be resolved before merging? it's just the name of the branch,I think...

vict0rsch commented:
@sashavor you're right there was a conflict because Tianyu commented out a bad import and I removed it so it was conflicting but I just deleted it in the end
